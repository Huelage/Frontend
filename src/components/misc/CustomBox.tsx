import { useAppTheme } from "@hooks";
import { Canvas, RoundedRect, Shadow } from "@shopify/react-native-skia";
import React, { memo } from "react";
import { StyleSheet } from "react-native";

interface CustomBoxInterface {
  height: number;
  width: number;
  pad: number;
  r: number;
  left?: number;
  bgColor?: string;
  shadowColor?: string;
}

const CustomBox = ({ height, width, pad, r, left, bgColor, shadowColor }: CustomBoxInterface) => {
  const { color } = useAppTheme();
  return (
    <Canvas style={[styles.container, { height, width, left: left || 0 }]}>
      <RoundedRect x={pad} y={pad} width={width - 20} height={height - 20} r={r} color={bgColor ?? color.cardBg}>
        <Shadow dx={4} dy={4} blur={4} color={shadowColor ?? color.cardShadow} />
      </RoundedRect>
    </Canvas>
  );
};

export default memo(CustomBox);

const styles = StyleSheet.create({
  container: {
    position: "absolute",
    top: 0
  }
});