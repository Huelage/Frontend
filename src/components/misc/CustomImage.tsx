import { Canvas, Fit, Group, Image, Shadow, SkRRect, rect, rrect, useImage } from '@shopify/react-native-skia';
import React from 'react';
import { StyleProp, ViewStyle } from 'react-native';

interface ImageInterface {
  imgUrl: string;
  imgSize: number;
  imgPad: number;
  imgFit?: Fit;
  shadowBlur?: number;
  shadowColor?: string;
  shadowHeight?: number;
  shadowWidth?: number;
  style: StyleProp<ViewStyle>;
}

const CustomImage = ({ imgUrl, imgSize, imgPad, imgFit, style, shadowBlur, shadowColor, shadowHeight, shadowWidth }: ImageInterface) => {
  const img = useImage(imgUrl);
  const roundedRect = { rect: { x: imgPad, y: imgPad, width: imgSize, height: imgSize }, rx: 1000, ry: 1000 };
  return (
    <Canvas style={style}>
      <Shadow dx={shadowWidth || 0} dy={shadowHeight || 6} blur={shadowBlur || 5} color={shadowColor || "rgba(0, 0, 0, .4)"} />
      <Group clip={roundedRect}>
        <Image image={img} x={imgPad} y={imgPad} width={imgSize} height={imgSize} fit={imgFit || "cover"} />
      </Group>
    </Canvas>
  );
};

export default CustomImage;