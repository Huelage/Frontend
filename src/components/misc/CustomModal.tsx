import { BlurView } from 'expo-blur';
import React from 'react';
import { StyleSheet } from 'react-native';
import { TouchableWithoutFeedback } from 'react-native-gesture-handler';
import RNModal from 'react-native-modal';
import { heightPercentageToDP as hp, widthPercentageToDP as wp } from 'react-native-responsive-screen';

interface ModalProps {
  isVisible: boolean;
  children: React.ReactNode;
  close: React.Dispatch<React.SetStateAction<boolean>>;
  [x: string]: any;
}

const CustomModal = ({ isVisible, children, close, ...props }: ModalProps) => {
  return (
    <RNModal
      isVisible={isVisible}
      animationIn="zoomIn"
      animationOut="zoomOut"
      backdropTransitionInTiming={500}
      backdropTransitionOutTiming={500}
      style={styles.modal}
    >
      <TouchableWithoutFeedback onPress={() => close(false)}>
        <BlurView intensity={9.5} style={styles.modalContainer}>
          <TouchableWithoutFeedback>
            {children}
          </TouchableWithoutFeedback>
        </BlurView>
      </TouchableWithoutFeedback>
    </RNModal>
  );
};

export default CustomModal;

const styles = StyleSheet.create({
  modal: {
    margin: 0
  },
  modalContainer: {
    alignItems: 'center',
    height: hp('100%'),
    justifyContent: 'center',
    width: wp('100%'),
  },
  childrenContainer: {

  },
  blurView: {
    position: 'absolute',
    top: 0,
    bottom: 0,
    left: 0,
    right: 0,
    backgroundColor: 'transparent'
  }
});