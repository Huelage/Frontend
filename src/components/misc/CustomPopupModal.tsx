import { useAppTheme } from '@hooks';
import React from 'react';
import { StyleSheet, TouchableOpacity, View } from 'react-native';
import RNModal from 'react-native-modal';
import { widthPercentageToDP as wp } from 'react-native-responsive-screen';
import { useSafeAreaInsets } from 'react-native-safe-area-context';

interface CustomModalProps {
  children: React.ReactNode;
  showModal: boolean;
  setShowModal: React.Dispatch<React.SetStateAction<boolean>>;
}

const CustomPopupModal = ({ children, showModal, setShowModal }: CustomModalProps) => {
  const { color } = useAppTheme();
  const insets = useSafeAreaInsets();
  return (
    <RNModal
      isVisible={showModal}
      animationIn="slideInUp"
      animationOut="slideOutDown"
      animationOutTiming={1000}
      style={styles.modal}
      testID='custom popup modal'
    >
      <View style={[styles.filterBox, { backgroundColor: color.cardBg2, paddingBottom: insets.bottom }]}>
        <TouchableOpacity onPress={() => setShowModal(false)} style={styles.boxCloser} testID='modal closer'>
          <View style={styles.boxCloserButton} />
        </TouchableOpacity>
        {children}
      </View>
    </RNModal>
  );
};

export default CustomPopupModal;

const styles = StyleSheet.create({
  modal: {
    margin: 0
  },
  filterBox: {
    alignItems: 'center',
    borderRadius: 30,
    bottom: 0,
    padding: 15,
    position: 'absolute',
    width: wp('100%'),
  },
  boxCloser: {
    paddingVertical: 10
  },
  boxCloserButton: {
    alignSelf: 'center',
    backgroundColor: "#BCB5B5",
    borderRadius: 5,
    height: 5,
    marginBottom: 10,
    width: 50
  }
});