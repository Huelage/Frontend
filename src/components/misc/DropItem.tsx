import { FontAwesome } from "@expo/vector-icons";
import { useAppTheme } from "@hooks";
import { DropDataInterface } from "@interfaces";
import { fonts, shadowStyle } from "@utils";
import React from "react";
import { Image, StyleSheet, Text, TouchableOpacity, View } from "react-native";
import Animated, { FadeInUp, ZoomOut } from "react-native-reanimated";
import { widthPercentageToDP as wp } from "react-native-responsive-screen";

interface DropItemInterface {
  idx: number;
  item: DropDataInterface;
  totalItems: number;
}

const DropItem = ({ item, idx, totalItems }: DropItemInterface) => {
  const { color } = useAppTheme();
  return (
    <Animated.View
      entering={FadeInUp.delay(100 + idx * 100)}
      exiting={ZoomOut.delay(100 + (totalItems - idx) * 100)}
      style={[styles.optionItem, { backgroundColor: color.main }]}
    >
      <View style={[styles.optionImageBox, { backgroundColor: color.main }]}>
        <Image source={{ uri: item.imgUrl }} style={styles.optionImage} />
      </View>
      <View style={styles.infoBox}>
        <Text style={[styles.headerText, { color: color.mainText }]}>{item.value}</Text>
        {item.desc ? <Text style={[styles.descText, { color: color.mainTextDim }]}>{item.desc}</Text> : null}
      </View>
      <TouchableOpacity style={[styles.iconBox, { backgroundColor: color.main }]}>
        <FontAwesome name="angle-right" size={24} color={color.mainGreen} style={{ marginLeft: 2 }} />
      </TouchableOpacity>
    </Animated.View>
  );
};

export default DropItem;

const styles = StyleSheet.create({
  optionItem: {
    alignItems: "center",
    borderRadius: 10,
    flexDirection: "row",
    height: 80,
    left: 30,
    marginBottom: 15,
    padding: 20,
    width: wp("100%") - 110,
    ...shadowStyle
  },
  optionImageBox: {
    alignItems: "center",
    borderRadius: 20,
    height: 60,
    justifyContent: "center",
    left: -30,
    padding: 10,
    position: "absolute",
    width: 60,
    ...shadowStyle
  },
  optionImage: {
    borderRadius: 25,
    height: 50,
    width: 50
  },
  infoBox: {
    flex: 1,
    justifyContent: "center",
    marginHorizontal: 20
  },
  headerText: {
    fontFamily: fonts.I_700,
    fontSize: 20,
  },
  descText: {
    fontFamily: fonts.I_300,
    fontSize: 12
  },
  iconBox: {
    alignItems: "center",
    borderRadius: 20,
    height: 40,
    justifyContent: "center",
    position: "absolute",
    right: -20,
    width: 40,
    ...shadowStyle
  }
});