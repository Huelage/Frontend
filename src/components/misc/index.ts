export { default as CustomBox } from "./CustomBox";
export { default as CustomDropdown } from "./CustomDropdown";
export { default as CustomFilterBox } from "./CustomFilterBox";
export { default as CustomImage } from "./CustomImage";
export { default as CustomModal } from "./CustomModal";
export { default as CustomPopupModal } from "./CustomPopupModal";
export { default as StarRating } from "./StarRating";
