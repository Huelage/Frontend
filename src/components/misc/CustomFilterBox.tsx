import { FontAwesome, Ionicons } from "@expo/vector-icons";
import { useAppTheme } from "@hooks";
import { FilterGroup, FilterItem } from "@interfaces";
import { fonts } from "@utils";
import React, { useCallback, useEffect, useMemo, useState } from "react";
import { FlatList, StyleSheet, Text, TouchableOpacity, View } from "react-native";
import { widthPercentageToDP as wp } from "react-native-responsive-screen";
import CustomBox from "./CustomBox";
import CustomPopupModal from "./CustomPopupModal";

interface CustomFilterBoxInterface {
  filterItems: FilterGroup[];
  defaultFilter: string;
}

const CustomFilterBox = ({ filterItems, defaultFilter }: CustomFilterBoxInterface) => {
  const { color } = useAppTheme();
  const [currentGroup, setCurrentGroup] = useState<string>(defaultFilter);
  const [currentGroupItems, setCurrentGroupItems] = useState<FilterItem[]>([]);
  const [isFilterBoxVisible, setIsFilterBoxVisible] = useState<boolean>(false);
  const [checkedItems, setCheckedItems] = useState<{ group: string, id: string; }[]>([]);

  const groups = useMemo(() => filterItems.map(item => item.label), [filterItems]);
  const groupTypeMapping = useMemo(() => {
    const mapping: Record<string, string> = {};
    filterItems.forEach(item => mapping[item.label] = item.type);
    return mapping;
  }, [filterItems]);

  const isChecked = useCallback((itemId: string) => !!checkedItems.find(check => check.id === itemId), [checkedItems]);
  const handleCheck = useCallback((item: FilterItem) => {
    if (isChecked(item.id)) {
      setCheckedItems(checkedItems.filter(check => check.id !== item.id));
      item.onPress(false);
    } else {
      let newCheckedItems: { group: string, id: string; }[] = checkedItems;
      if (groupTypeMapping[currentGroup] === "SINGLE")
        newCheckedItems = checkedItems.filter(check => check.group != item.groupId);
      setCheckedItems([...newCheckedItems, { id: item.id, group: item.groupId }]);
      item.onPress(true);
    }
  }, [groupTypeMapping, checkedItems, currentGroup]);

  useEffect(() => {
    setCurrentGroupItems(filterItems.find(item => item.label === currentGroup)?.items || []);
  }, [currentGroup]);
  return (
    <View testID="custom filter box">
      <TouchableOpacity style={styles.container} onPress={() => setIsFilterBoxVisible(!isFilterBoxVisible)} testID="toggle filter box">
        <CustomBox bgColor={color.cardBg2} height={50} r={50} pad={6} width={120} left={-6} />
        <Text style={[styles.titleText, { color: color.mainText }]}>Filter by</Text>
        <FontAwesome name="angle-down" size={18} color={color.mainText} />
      </TouchableOpacity>
      <CustomPopupModal setShowModal={setIsFilterBoxVisible} showModal={isFilterBoxVisible}>
        <View style={styles.filterBox}>
          <FlatList
            data={groups}
            horizontal
            contentContainerStyle={styles.filterHeader}
            keyExtractor={(item) => item}
            renderItem={({ item }) => (
              <TouchableOpacity onPress={() => setCurrentGroup(item)} style={[styles.filterHeaderButton, { borderColor: item === currentGroup ? color.mainGreen : "transparent" }]} testID="filter header item">
                <Text style={[styles.filterHeaderText, { color: item === currentGroup ? color.mainGreen : color.mainText }]}>{item}</Text>
              </TouchableOpacity>
            )}
            testID="filter header list"
          />
          <FlatList
            data={currentGroupItems}
            ItemSeparatorComponent={() => <View style={styles.filterItemSeperator} />}
            keyExtractor={item => item.id}
            renderItem={({ item }) => (
              <TouchableOpacity onPress={() => handleCheck(item)} testID="filter item">
                <View style={styles.filterItem}>
                  <Text style={[styles.filterItemText, { color: isChecked(item.id) ? color.mainGreen : color.mainText }]}>{item.name}</Text>
                  {isChecked(item.id) ? <Ionicons name="checkmark" size={24} color={color.mainGreen} testID={`${item.name} checkmark`} /> : null}
                </View>
              </TouchableOpacity>
            )}
            testID="filter items list"
          />
        </View>
      </CustomPopupModal>
    </View>
  );
};

export default CustomFilterBox;

const styles = StyleSheet.create({
  container: {
    alignItems: "center",
    borderRadius: 50,
    flexDirection: "row",
    gap: 5,
    justifyContent: "space-between",
    paddingHorizontal: 15,
    paddingVertical: 12,
    width: 110
  },
  titleText: {
    fontFamily: fonts.I_600,
    fontSize: 14
  },
  filterBox: {
    alignItems: "center",
    borderRadius: 20,
    padding: 20,
    width: wp("100%")
  },
  filterHeader: {
    alignItems: "center",
    borderColor: "#BCB5B5",
    borderBottomWidth: 1,
    flexDirection: "row",
    gap: 20,
    marginBottom: 10,
    width: "100%"
  },
  filterHeaderButton: {
    borderBottomWidth: 2,
    paddingBottom: 10,
  },
  filterHeaderText: {
    fontFamily: fonts.I_700,
    fontSize: 16,
    letterSpacing: 1,
    textTransform: "capitalize"
  },
  filterItem: {
    alignItems: "center",
    flex: 1,
    flexDirection: "row",
    justifyContent: "space-between",
    paddingVertical: 10,
    height: 45,
    width: wp("100%") - 60,
  },
  filterItemSeperator: {
    height: 1,
    backgroundColor: "#BCB5B5"
  },
  filterItemText: {
    fontFamily: fonts.I_400
  }
});