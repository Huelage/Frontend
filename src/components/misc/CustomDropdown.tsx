import { useAppTheme } from "@hooks";
import { DropDataInterface } from "@interfaces";
import { fonts } from "@utils";
import React, { useEffect, useState } from "react";
import { StyleSheet, Text, TouchableOpacity, View } from "react-native";
import Animated, { Layout, interpolate, useAnimatedStyle, useSharedValue, withTiming } from "react-native-reanimated";
import DropItem from "./DropItem";

interface CustomDropdownProps {
  data: DropDataInterface[];
  label: string,
  isError?: boolean;
  value?: string;
  onChange: (val: string) => void;
}

const CustomDropdown = ({ data, label, isError, value, onChange }: CustomDropdownProps) => {
  const { color } = useAppTheme();
  const dropHeight = useSharedValue(0);
  const [isDropOpen, setIsDropOpen] = useState<boolean>(false);
  const [selected, setSelected] = useState("");

  const animatedDropHeight = useAnimatedStyle(() => ({
    height: interpolate(dropHeight.value, [0, 1], [0, 96 * data.length])
  }));
  const onSelect = (value: string) => {
    setSelected(value);
    onChange(value);
    setIsDropOpen(false);
  };

  useEffect(() => {
    dropHeight.value = withTiming(isDropOpen ? 1 : 0, { duration: 200 * data.length });
  }, [data, isDropOpen]);
  useEffect(() => {
    setSelected(value || "");
  }, [value]);
  return (
    <View style={styles.dropContainer} testID="custom dropdown">
      <Text style={[styles.dropLabel, { color: color.mainText }]}>{label}</Text>
      <TouchableOpacity onPress={() => setIsDropOpen(!isDropOpen)} style={[styles.mainBox, { borderColor: isError ? color.danger : color.mainGreen }]} testID="dropdown toggle">
        <Text style={{ color: selected ? color.mainText : color.mainTextDim }}>{selected ? selected : "Select options..."}</Text>
      </TouchableOpacity>
      <Animated.FlatList
        data={isDropOpen ? data : []}
        keyExtractor={(item, idx) => `${item.value}-${idx}`}
        itemLayoutAnimation={Layout.springify().damping(15).delay(150)}
        scrollEnabled={false}
        renderItem={({ item, index }) => (
          <TouchableOpacity onPress={() => onSelect(item.value)} testID={`dropdown item ${item.value}`}>
            <DropItem item={item} idx={index} totalItems={data.length - 1} />
          </TouchableOpacity>
        )}
        contentContainerStyle={[styles.optionBox]}
        style={animatedDropHeight}
      />
    </View>
  );
};

export default CustomDropdown;

const styles = StyleSheet.create({
  dropContainer: {
    gap: 5
  },
  dropLabel: {
    fontFamily: fonts.I_600,
    fontSize: 16
  },
  mainBox: {
    borderRadius: 10,
    borderWidth: 2,
    height: 40,
    justifyContent: "center",
    paddingHorizontal: 15
  },
  dropInput: {
    fontFamily: fonts.I_400,
    fontSize: 16
  },
  optionBox: {
    justifyContent: "center",
    paddingHorizontal: 10,
    paddingTop: 5
  },

});