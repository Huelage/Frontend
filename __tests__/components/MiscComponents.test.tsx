import { CustomDropdown, CustomFilterBox, CustomModal, CustomPopupModal, StarRating } from "@components/misc";
import { FilterGroup } from "@interfaces";
import { fireEvent, render, screen, waitFor } from "@testing-library/react-native";
import { Text } from "react-native";

describe("When Testing Miscellaneous Components: ", () => {
  describe("<CustomDropdown />: ", () => {
    const data = [
      { value: "testValue1", imgUrl: "image", desc: "test description 1" },
      { value: "testValue2", imgUrl: "image" },
      { value: "testValue3", imgUrl: "image", desc: "test description 3" }
    ];
    const onChange = jest.fn();
    beforeEach(() => {
      render(<CustomDropdown data={data} label="Test dropdown" onChange={onChange} />);
    });
    it("should render the component correctly", () => {
      expect(screen.getByTestId("custom dropdown")).toBeOnTheScreen();
    });
    it("should render the label", () => {
      expect(screen.getByText("Test dropdown")).toBeOnTheScreen();
    });
    it("should render the dropdown", () => {
      render(<CustomDropdown data={data} label="Test dropdown" onChange={onChange} isError />);
      expect(screen.getByTestId("dropdown toggle")).toBeOnTheScreen();
    });
    it("should call the onChange function when the dropdown is used", () => {
      fireEvent.press(screen.getByTestId("dropdown toggle"));
      fireEvent.press(screen.getByTestId("dropdown item testValue2"));
      expect(onChange).toBeCalled();
    });
  });

  describe("<CustomFilterBox />: ", () => {
    const onPress = jest.fn();
    const props = {
      filterItems: [
        {
          id: "status", label: "Status", type: "MULTIPLE", items: [
            { id: "1", groupId: "status", name: "Pending", onPress },
            { id: "2", groupId: "status", name: "Completed", onPress }
          ]
        },
        {
          id: "date", label: "Order date", type: "SINGLE", items: [
            { id: "4", groupId: "date", name: "Today", onPress },
            { id: "5", groupId: "date", name: "Yesterday", onPress }
          ]
        }
      ] as FilterGroup[], defaultFilter: "Status"
    };

    beforeEach(() => {
      render(<CustomFilterBox {...props} />);
    });
    // Testing UI
    it("should render the component correctly", () => {
      expect(screen.getByTestId("custom filter box")).toBeOnTheScreen();
    });
    it("should render the toggle filter box button", () => {
      expect(screen.getByTestId("toggle filter box")).toBeOnTheScreen();
    });
    it("should render the CustomPopupModal component", () => {
      expect(screen.getByTestId("custom popup modal")).toBeOnTheScreen();
    });
    it("should render the filter header list", () => {
      expect(screen.getByTestId("filter header list")).toBeOnTheScreen();
    });
    it("should render the filter header items", () => {
      expect(screen.getAllByTestId("filter header item")).toHaveLength(2);
    });
    it("should render the filter items list", () => {
      expect(screen.getByTestId("filter items list")).toBeOnTheScreen();
    });
    it("should render the filter items", () => {
      expect(screen.getAllByTestId("filter item")).toHaveLength(2);
    });
    it("should render an empty list when filter items are not available", () => {
      render(<CustomFilterBox {...props} filterItems={[]} />);
      expect(screen.queryAllByTestId("filter item")).toHaveLength(0);
    });
    // Testing Functionality
    it("should open the filter box when the toggle filter box button is pressed", () => {
      const filterBox = screen.getByTestId("custom popup modal");
      const toggleFilterBox = screen.getByTestId("toggle filter box");
      expect(filterBox.props.isVisible).toBeFalsy();
      fireEvent.press(toggleFilterBox);
      expect(filterBox.props.isVisible).toBeTruthy();
    });
    it("should select a new group when one of the header items is pressed", async () => {
      const filterHeaderItem = screen.getAllByTestId("filter header item")[1];
      expect(screen.queryByText("Yesterday")).toBeNull();
      fireEvent.press(filterHeaderItem);
      await waitFor(() => expect(screen.getByText("Yesterday")).toBeOnTheScreen());
    });
    it("should select a new filter item when one of the filter items is pressed", async () => {
      const filterItem = screen.getAllByTestId("filter item")[1];
      expect(screen.queryByText("Completed checkmark")).toBeNull();
      fireEvent.press(filterItem);
      await waitFor(() => expect(screen.getByTestId("Completed checkmark")).toBeOnTheScreen());
      expect(onPress).toBeCalledWith(true);
    });
    it("should replace all items from a group and select the new filter if the group type is SINGLE", async () => {
      const filterHeaderItem = screen.getAllByTestId("filter header item")[1];
      fireEvent.press(filterHeaderItem);
      expect(screen.queryByText("Yesterday checkmark")).toBeNull();
      fireEvent.press(screen.getAllByTestId("filter item")[1]);
      await waitFor(() => expect(screen.getByTestId("Yesterday checkmark")).toBeOnTheScreen());
      fireEvent.press(screen.getAllByTestId("filter item")[0]);
      await waitFor(() => expect(screen.getByTestId("Today checkmark")).toBeOnTheScreen());
      expect(onPress).toBeCalledWith(true);
    });
    it("should unselect a filter item when it is pressed again", async () => {
      const filterItem = screen.getAllByTestId("filter item")[1];
      expect(screen.queryByText("Completed checkmark")).toBeNull();
      fireEvent.press(filterItem);
      await waitFor(() => expect(screen.getByTestId("Completed checkmark")).toBeOnTheScreen());
      fireEvent.press(filterItem);
      await waitFor(() => expect(screen.queryByText("Completed checkmark")).toBeNull());
      expect(onPress).toBeCalledWith(false);
    });
  });

  describe("<CustomModal />: ", () => {
    beforeEach(() => {
      render(
        <CustomModal isVisible={true}>
          <Text>test modal content</Text>
        </CustomModal>
      );
    });
    it("should render correctly", () => {
      expect(screen.getByTestId("custom modal")).toBeOnTheScreen();
    });
    it("should render the children correctly", () => {
      expect(screen.getByText("test modal content")).toBeOnTheScreen();
    });
  });

  describe("<CustomPopupModal />: ", () => {
    const handleModal = jest.fn();
    beforeEach(() => {
      render(
        <CustomPopupModal showModal={true} setShowModal={handleModal}>
          <Text>test modal content</Text>
        </CustomPopupModal>
      );
    });
    it("should render correctly", () => {
      expect(screen.getByTestId("custom popup modal")).toBeOnTheScreen();
    });
    it("should render the children correctly", () => {
      expect(screen.getByText("test modal content")).toBeOnTheScreen();
    });
    it("should close the modal when the modal closer bar is pressed", () => {
      const modalCloser = screen.getByTestId("modal closer");
      fireEvent.press(modalCloser);
      expect(handleModal).toBeCalledWith(false);
    });
  });

  describe("<StarRating />: ", () => {
    it("should render correctly", () => {
      render(<StarRating rating={3} />);
      expect(screen.getByTestId("star rating")).toBeOnTheScreen();
    });
    it("should render the correct number of full stars", () => {
      render(<StarRating rating={3} />);
      expect(screen.getAllByTestId("ios-star")).toHaveLength(3);
    });
    it("should render the correct number of half stars", () => {
      render(<StarRating rating={3.5} />);
      expect(screen.getAllByTestId("ios-star-half")).toHaveLength(1);
    });
    it("should render the correct number of empty stars", () => {
      render(<StarRating rating={3.5} />);
      expect(screen.getAllByTestId("ios-star-outline")).toHaveLength(1);
    });
  });
});
