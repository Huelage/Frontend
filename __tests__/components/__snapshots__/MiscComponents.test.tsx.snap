// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`When Testing Miscellaneous Components:  <StarRating />:  should render correctly 1`] = `
<View
  style={
    [
      {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "center",
      },
      {
        "gap": 3,
      },
    ]
  }
  testID="star rating"
>
  <Text />
  <Text />
  <Text />
  <Text />
  <Text />
</View>
`;
